README.MD
# Small C repository

Collected at one place for historic purposes.

This is original untouched source code.

* ASM - Small Assembler 1.2 (rev level 14)
* C - Small C 2.2 (rev level 117)
* TOOLS - Small Tools 1.2 (rev level 33)
* VAL - Linker, released in public domain at SDC'89
* SW - Small Windows 1.1 (rev level 18) for Small C
* SW-TC - Small Windows 1.1 (rev level 18) for Turbo C (any version)

Assembler, C and VAL could be used together.

MAKE tool for Small-C available from https://github.com/DosWorld/smallmake

# License

I am not a author of this source codes and nothing is contributed by me. Here is original Licenses for each project:

## Small C, Small Assembler, Small Tools and Small Windows

Notice of Public Domain Status

The source code for the Small-C Compiler and runtime libraries (CP/M & DOS), Small-Mac Assembler (CP/M), Small-Assembler (DOS), Small-Tools programs and Small-Windows library to which I hold copyrights are hereby available for royalty free use in private or commerical endeavors. The only obligation being that the users retain the original copyright notices and credit all prior authors (Ron Cain, James Hendrix, etc.) in derivative versions.

James E. Hendrix Jr.

http://www.deturbulator.org/Jim.asp

## VAL Linker

This program is released to the public domain by the author on
February 15, 1989 at the Software Development Conference,
San Francisco, CA.  With the understanding that there are no
warranties of any sort either expressed or implied, you may use the
code in whole or part for any lawful purpose.  The author does not
warrant or guarantee that this code is fit for any particular purpose.
You are under no obligation to acknowledge the author when using this
code, but you may do so if you desire.  The author neither seeks nor
desires any contributions for this code.

Should you desire to communicate with the author, the preferred method
is via the author's bulletin board, Pontchippi.  Pontchippi is an
OPUS board.  You may contact it directly at 504-244-1417 or via net
mail to node 390/105.  The board has a 9600 baud US Robotics Courier
HST modem.  Should you contact the board directly, access to the board
is granted immediately up completion of a very simple questionnaire.

The author wishes to acknowledge and thank Jack Purdum and Tim Leslie
of Ecosoft for their assistance and support.  In addition to Jack and
Tim, Steve Russell of SLR Systems was also of invaluable assistance.
All helped in this project by debugging and offering their expert
advice and guidance.  They always had time for me seemingly whenever
I requested it.

Finally, this program is dedicated to the memory of my father-in-law,
Mariano "Val" Valentino.  The only request the author makes for
using the program is that the program name be left as "VAL" in his
memory.

    David Troendle
    New Orleans, LA
    February 10, 1989


C/HISTORY
Small-C Revision History

 -- included in version 2.2

 76 Modified _read() in CSYSLIB.C to do console input through _getkey()
    for consistent handling of special key values per the MS-DOS addendum
    to the Small-C Handbook.
 77 Modified fgetc() in FGETC.C to look for an ungotten character and
    return it without cooking it (since it cooked on the original get).
    This fixed an apparent problem in scanf() and fscanf() in which
    they appear to miss input characters.
 78 Revised out_code() and do more optimizing.  Let p-code subscript code[],
    (an array of pointers to output assembly code strings) instead of using
    a monstrous switch statement.
 79 Always compile features that are now considered standard.
 80 Properly calculate swend in main() to use all of the switch table.
 81 Dropped unnecessary variable func1.
 82 Changed if(match(...)); to match(...); statements.
 83 Dropped unnecessary while loop in declloc().
 84 Simplified treatment of argument "dim" in initials().
 85 Dropped outside() because it is redundant with code in compound().
 86 Revised the point at which lval2[] is declared in plnge(), and hier14()
    to greatly reduce the amount of stack space used during expression
    analysis by allocating lval[] arrays only when actually needed. 
 87 Simplified logic in newfunc(), doif(), dodo(), preprocess(), skip(), 
    plnge(), expression(), hier14(), primary(), pstr(), white().  
 88 Modified primary() to add an undeclared function to the symbol table only 
    if the name is not already in the table.  This improves run time by 
    saving a table lookup on most function references.  
 89 Reorganized source code into four files (CC1.C, CC2.C, CC3.C, CC4.C),
    regrouped functions into more logical units, and renamed some functions
    (addmac->dodefine, newfunc->dofunction, rvalue->fetch, pstr->chrcon,
    qstr->string, needtoken->need, col->colon, ol->outline, junk->skip).
 90 Improved output efficiency by deleting cout(), sout(), and xout().
    It now checks for an output error only at the end of a line.
 91 Dropped assignment statements in main() that set globals to zero.
 92 Revised ispunct() to exclude white space from the 'true' condition.
 93 Combined all of the is...() character classification functions into a
    single module of short, fast functions that subscript into an array
    for the answer.  This approach is three times faster.
 94 Per Kiyoshi Akima, eliminated LINK module from library.  Now, trailer()
    declares _MAIN to be external if main() is declared in the module being 
    compiled; this forces CSYSLIB to load when linking.  By not declaring it 
    external in other modules, a "multiply defined" error is eliminated when 
    compiling CSYSLIB which contains _MAIN.  However, since .OBJ modules
    which were compiled with earlier versions of Small-C contain references to 
    __LINK, it is declared as a dummy entry point in CALL.ASM to avoid 
    "unresolved reference" errors at link time.
 95 Per Kiyoshi Akima, installed additional code optimizing.
 96 Dropped setops() in favor of initializing the op[] and op2[] arrays.
    This can be done since these arrays now contain p-codes rather than
    function addresses.
 97 Yet another rewrite of the the optimizer.  It's now much smaller, each 
    optimization is cheaper, the logic is clearer and more bullet proof.
    And, yet more optimizations.  EXE sizes are down by 10% and execution times
    are down by as much as 25%.
 98 Renamed CC.DEF to CC.H and CLIB.DEF to CLIB.H.
 99 Corrected unlink() to return ERR on failure instead of EOF.
100 Revision 76 slowed down the reading of files considerably by introducing
    an additional call to iscons() for every character read.  It was discovered
    that the DOS interrupt in this function is very very slow.  So, iscons()
    was revised to call DOS only the first time for a file, and then test
    the rememberred answer thereafter.  This required changes to CSYSLIB.C
    also.  The effect was a DOUBLING OF I/O SPEED DURING READS, compared to
    revision level 75 -- BEFORE the slowdown was introduced.  This shows that
    the existing calls to iscons() in fgetc() and fgets() have been a major
    bottleneck all along.
101 Added the "unsigned" integer and character data types.  These result in
    unsigned <, <=, >, >=, *, /, % operations and yield unsigned results.
102 Revised the handling of numeric constants to recognize values from
    32768 to 65535 as unsigned.  These, likewise, produce unsigned operations
    and yield unsigned results.
103 Cleaned up the handling of non-definitional function declarations.  The 
    syntax (*func)(), which specifies a pointer to a function, is now accepted 
    only for function arguments.  Previously, it was accepted at the global and 
    local levels, but not interpreted as a pointer.
104 Reject instances of declarations, such as *func(), which have multiple
    modifiers.  Previously, the right most modifier prevailed.  But it is 
    better to reject something than to interpret in differently than it is
    written.
105 Accept "void" before function header as comment.
106 Corrected an error in the expression analyzer which causes pointer offsets
    to be lost when subscripting is applied; e.g., (ptr+5)[x] would lose the
    offset 5.
107 Revised the implementation of the sizeof() operator so it works correctly.
    Although undocumented, this was in the original MS-DOS compiler.
108 Always pass an argument count to functions, even if NOCCARGC is defined.
    This is not very costly on 8086 family CPUs and it eliminates a lot of
    headaches when you call printf() but forget that NOCCARGC is defined.
109 Small-C programs now abort with a return code of 1 for memory overflows
    and 2 for operator interruptions (control-C).
110 Revised strlen() to use the 8086 "REPNE SCASB" prefix/instruction for 

C/CC.H
/*
** CC.H -- Symbol Definitions for Small-C compiler.
*/

/*
** machine dependent parameters
*/
#define BPW     2   /* bytes per word */
#define LBPW    1   /* log2(BPW) */
#define SBPC    1   /* stack bytes per character */
#define ERRCODE 7   /* op sys return code */

/*
** symbol table format
*/
#define IDENT    0
#define TYPE     1
#define CLASS    2
#define SIZE     3
#define OFFSET   5
#define NAME     7

#define SYMAVG  12
#define SYMMAX  16

/*
** symbol table parameters
*/
#define NUMLOCS   25
#define STARTLOC  symtab
#define ENDLOC    (symtab+NUMLOCS*SYMAVG)
#define NUMGLBS   200
#define STARTGLB  ENDLOC
#define ENDGLB    (ENDLOC+(NUMGLBS-1)*SYMMAX)
#define SYMTBSZ   3050  /* (NUMLOCS*SYMAVG + NUMGLBS*SYMMAX) */

/*
** system wide name size (for symbols)
*/
#define NAMESIZE 9
#define NAMEMAX  8
 
/*
** values for "IDENT"
*/
#define LABEL    0
#define VARIABLE 1
#define ARRAY    2
#define POINTER  3
#define FUNCTION 4
 
/*
** values for "TYPE"
**    high order 14 bits give length of object
**    low order 2 bits make type unique within length
*/
/*      LABEL   0 */
#define CHR     (  1 << 2)
#define INT     (BPW << 2)
#define UCHR   ((  1 << 2) + 1)
#define UINT   ((BPW << 2) + 1)
#define UNSIGNED             1
 
/*
** values for "CLASS"
*/
/*      LABEL     0 */
#define AUTOMATIC 1
#define STATIC    2
#define EXTERNAL  3
#define AUTOEXT   4

/*
** segment types
*/
#define DATASEG 1
#define CODESEG 2

/*
** "switch" table
*/
#define SWSIZ   (2*BPW)
#define SWTABSZ (90*SWSIZ)
 
/*
** "while" queue
*/
#define WQTABSZ  30
#define WQSIZ     3
#define WQMAX   (wq+WQTABSZ-WQSIZ)
 
/*
** field offsets in "while" queue
*/
#define WQSP    0
#define WQLOOP  1
#define WQEXIT  2
 
/*
** literal pool
*/
#define LITABSZ 2000
#define LITMAX  (LITABSZ-1)
 
/*
** input line
*/
#define LINEMAX  127
#define LINESIZE 128

/*
** entries in staging buffer
*/
#define STAGESIZE   200
 
/*
** macro (#define) pool
*/
#define MACNBR   300
#define MACNSIZE (MACNBR*(NAMESIZE+2))
#define MACNEND  (macn+MACNSIZE)
#define MACQSIZE (MACNBR*7)
#define MACMAX   (MACQSIZE-1)
 
/*
** statement types
*/
#define STIF      1
#define STWHILE   2
#define STRETURN  3
#define STBREAK   4
#define STCONT    5
#define STASM     6
#define STEXPR    7
#define STDO      8
#define STFOR     9
#define STSWITCH 10
#define STCASE   11
#define STDEF    12
#define STGOTO   13
#define STLABEL  14

/*
** p-code symbols
**
** legend:
**  1 = primary register (pr in comments)
**  2 = secondary register (sr in comments)
**  b = byte
**  f = jump on false condition
**  l = current literal pool label number
**  m = memory reference by label
**  n = numeric constant
**  p = indirect reference thru pointer in sr
**  r = repeated r times
**  s = stack frame reference
**  u = unsigned
**  w = word
**  _ (tail) = another p-code completes this one
*/

        /* compiler-generated */
#define ADD12     1   /* add sr to pr */
#define ADDSP     2   /* add to stack pointer */
#define AND12     3   /* AND sr to pr */
#define ANEG1     4   /* arith negate pr */
#define ARGCNTn   5   /* pass arg count to function */
#define ASL12     6   /* arith shift left sr by pr into pr */
#define ASR12     7   /* arith shift right sr by pr into pr */
#define CALL1     8   /* call function thru pr */
#define CALLm     9   /* call function directly */
#define BYTE_    10   /* define bytes (part 1) */
#define BYTEn    11   /* define byte of value n */
#define BYTEr0   12   /* define r bytes of value 0 */
#define COM1     13   /* ones complement pr */
#define DBL1     14   /* double pr */
#define DBL2     15   /* double sr */
#define DIV12    16   /* div pr by sr */
#define DIV12u   17   /* div pr by sr unsigned */
#define ENTER    18   /* set stack frame on function entry */
#define EQ10f    19   /* jump if (pr == 0) is false */
#define EQ12     20   /* set pr TRUE if (sr == pr) */
#define GE10f    21   /* jump if (pr >= 0) is false */
#define GE12     22   /* set pr TRUE if (sr >= pr) */
#define GE12u    23   /* set pr TRUE if (sr >= pr) unsigned */
#define POINT1l  24   /* point pr to function's literal pool */
#define POINT1m  25   /* point pr to mem item thru label */
#define GETb1m   26   /* get byte into pr from mem thru label */
#define GETb1mu  27   /* get unsigned byte into pr from mem thru label */
#define GETb1p   28   /* get byte into pr from mem thru sr ptr */
#define GETb1pu  29   /* get unsigned byte into pr from mem thru sr ptr */
#define GETw1m   30   /* get word into pr from mem thru label */
#define GETw1n   31   /* get word of value n into pr */
#define GETw1p   32   /* get word into pr from mem thru sr ptr */
#define GETw2n   33   /* get word of value n into sr */
#define GT10f    34   /* jump if (pr > 0) is false */
#define GT12     35   /* set pr TRUE if (sr > pr) */
#define GT12u    36   /* set pr TRUE if (sr > pr) unsigned */
#define WORD_    37   /* define word (part 1) */
#define WORDn    38   /* define word of value n */
#define WORDr0   39   /* define r words of value 0 */
#define JMPm     40   /* jump to label */
#define LABm     41   /* define label m */
#define LE10f    42   /* jump if (pr <= 0) is false */
#define LE12     43   /* set pr TRUE if (sr <= pr) */
#define LE12u    44   /* set pr TRUE if (sr <= pr) unsigned */
#define LNEG1    45   /* logical negate pr */
#define LT10f    46   /* jump if (pr < 0) is false */
#define LT12     47   /* set pr TRUE if (sr < pr) */
#define LT12u    48   /* set pr TRUE if (sr < pr) unsigned */
#define MOD12    49   /* modulo pr by sr */
#define MOD12u   50   /* modulo pr by sr unsigned */
#define MOVE21   51   /* move pr to sr */
#define MUL12    52   /* multiply pr by sr */
#define MUL12u   53   /* multiply pr by sr unsigned */
#define NE10f    54   /* jump if (pr != 0) is false */
#define NE12     55   /* set pr TRUE if (sr != pr) */
#define NEARm    56   /* define near pointer thru label */
#define OR12     57   /* OR sr onto pr */
#define POINT1s  58   /* point pr to stack item */
#define POP2     59   /* pop stack into sr */
#define PUSH1    60   /* push pr onto stack */
#define PUTbm1   61   /* put pr byte in mem thru label */
#define PUTbp1   62   /* put pr byte in mem thru sr ptr */
#define PUTwm1   63   /* put pr word in mem thru label */
#define PUTwp1   64   /* put pr word in mem thru sr ptr */
#define rDEC1    65   /* dec pr (may repeat) */
#define REFm     66   /* finish instruction with label */
#define RETURN   67   /* restore stack and return */
#define rINC1    68   /* inc pr (may repeat) */
#define SUB12    69   /* sub sr from pr */
#define SWAP12   70   /* swap pr and sr */
#define SWAP1s   71   /* swap pr and top of stack */
#define SWITCH   72   /* find switch case */
#define XOR12    73   /* XOR pr with sr */

        /* optimizer-generated */
#define ADD1n    74   /* add n to pr */
#define ADD21    75   /* add pr to sr */
#define ADD2n    76   /* add immediate to sr */
#define ADDbpn   77   /* add n to mem byte thru sr ptr */
#define ADDwpn   78   /* add n to mem word thru sr ptr */
#define ADDm_    79   /* add n to mem byte/word thru label (part 1) */
#define COMMAn   80   /* finish instruction with ,n */
#define DECbp    81   /* dec mem byte thru sr ptr */
#define DECwp    82   /* dec mem word thru sr ptr */
#define POINT2m  83   /* point sr to mem thru label */
#define POINT2m_ 84   /* point sr to mem thru label (part 1) */
#define GETb1s   85   /* get byte into pr from stack */
#define GETb1su  86   /* get unsigned byte into pr from stack */
#define GETw1m_  87   /* get word into pr from mem thru label (part 1) */
#define GETw1s   88   /* get word into pr from stack */
#define GETw2m   89   /* get word into sr from mem (label) */
#define GETw2p   90   /* get word into sr thru sr ptr */
#define GETw2s   91   /* get word into sr from stack */
#define INCbp    92   /* inc byte in mem thru sr ptr */
#define INCwp    93   /* inc word in mem thru sr ptr */
#define PLUSn    94   /* finish instruction with +n */
#define POINT2s  95   /* point sr to stack */
#define PUSH2    96   /* push sr */
#define PUSHm    97   /* push word from mem thru label */
#define PUSHp    98   /* push word from mem thru sr ptr */
#define PUSHs    99   /* push word from stack */
#define PUT_m_  100   /* put byte/word into mem thru label (part 1) */
#define rDEC2   101   /* dec sr (may repeat) */
#define rINC2   102   /* inc sr (may repeat) */
#define SUB_m_  103   /* sub from mem byte/word thru label (part 1) */
#define SUB1n   104   /* sub n from pr */
#define SUBbpn  105   /* sub n from mem byte thru sr ptr */
#define SUBwpn  106   /* sub n from mem word thru sr ptr */

#define PCODES  107   /* size of code[] */

